frontend/src/metabase/dashboard/actions/save.js

```diff
@@ -16,6 +16,28 @@ import { fetchDashboard } from "./data-fetching";
 export const SAVE_DASHBOARD_AND_CARDS =
   "metabase/dashboard/SAVE_DASHBOARD_AND_CARDS";
 
+function hasDashboardChanged(dashboard, dashboardBeforeEditing) {
+  const dashboardHasChanged = !_.isEqual(
+    { ...dashboard, ordered_cards: dashboard.ordered_cards.length },
+    {
+      ...dashboardBeforeEditing,
+      ordered_cards: dashboardBeforeEditing.ordered_cards.length,
+    },
+  );
+
+  if (dashboardHasChanged) {
+    return true;
+  }
+
+  // sometimes the cards objects change order but all the cards themselves are the same
+  // this should not trigger a save
+  const cardsHaveChanged =
+    _.difference(dashboard.ordered_cards, dashboardBeforeEditing.ordered_cards)
```

Comment:
I think we can do this check in a more performant way. `_.difference`([source code](https://underscorejs.org/docs/modules/difference.html))  is essentially a double for loop with no early return, we only care if just one element in `dashboard.ordered_cards` is not in `dashboardBeforeEditing.ordered_cards`.

We can do this in `O(n)` time if we convert one of the arrays into a set, then iterate through the other array, and return `true` upon finding the first element that is not in the set.
```
function haveCardsChanged(dashboardCards, dashboardBeforeEditingCards) {
  const cardSet = new set(dashboardBeforeEditingCards)
  return dashboardCards.some(card => !cardSet.has(card))
}

Reply:
will `cardSet.has()` check deep equality?